### PR TITLE
refactor: use elastic cloud

### PR DESCRIPTION
**Pull request by lionelB**
use external elasticsearch

### CODE REVIEW

**Review by douglasduteil — APPROVED** [review]

File: .gitlab-ci/stages/deploy.yml [hunk 1]
```diff
@@ -1,11 +1,54 @@
...
+    - kubectl get secret elasticsearch-cloud-data --namespace=cdtn-secret --export -o yaml |
+      kubectl apply --namespace=${K8S_NAMESPACE} -f -
+    - kubectl get secret elasticsearch-cloud-api --namespace=cdtn-secret --export -o yaml |
+      kubectl apply --namespace=${K8S_NAMESPACE} -f -
+    - kubectl get secret elastic-apm --namespace=cdtn-secret --export -o yaml |
+      kubectl apply --namespace=${K8S_NAMESPACE} -f -
+
+#
+
+.resolve_helm_render_args: &resolve_helm_render_args |
```
Comment (douglasduteil) [inline .gitlab-ci/stages/deploy.yml]: to remove in v3 gitlab-ci-yaml
`PRODUCTION` var must be defined

File: .gitlab-ci/stages/deploy.yml [hunk 2]
```diff
@@ -1,11 +1,54 @@
...
 include:
-  - /.k8s/api/.k8s-deploy-api.yml
-  - /.k8s/data/.k8s-deploy-data.yml
-  - /.k8s/elasticsearch/.k8s-deploy-elasticsearch.yml
-  - /.k8s/frontend/.k8s-deploy-frontend.yml
-  - /.k8s/nlp/.k8s-deploy-nlp.yml
+  #
+  - "https://raw.githubusercontent.com/SocialGouv/gitlab-ci-yml/v2.2.1/base_create_namespace_stage.yml"
+  - "https://raw.githubusercontent.com/SocialGouv/gitlab-ci-yml/v2.2.1/base_deploy_nodejs_chart_stage.yml"
+  - "https://raw.githubusercontent.com/SocialGouv/gitlab-ci-yml/feat/hpa/base_deploy_hpa_chart_stage.yml"
```
Comment (douglasduteil) [inline .gitlab-ci/stages/deploy.yml]: Use merged version

File: .gitlab-ci/stages/deploy.yml [hunk 3]
```diff
@@ -1,11 +1,54 @@
...
+  - "https://raw.githubusercontent.com/SocialGouv/gitlab-ci-yml/v2.2.1/base_docker_helm_image_stage.yml"
+  - "https://raw.githubusercontent.com/SocialGouv/gitlab-ci-yml/v2.2.1/base_docker_kubectl_image_stage.yml"
+  - "https://raw.githubusercontent.com/SocialGouv/gitlab-ci-yml/v2.2.1/base_semantic_release_stage.yml"
+  - "https://raw.githubusercontent.com/SocialGouv/gitlab-ci-yml/v2.2.1/register-stage.yml"
```
Comment (douglasduteil) [inline .gitlab-ci/stages/deploy.yml]: Update to v3

File: .gitlab-ci/stages/deploy.yml [hunk 4]
```diff
@@ -16,11 +59,15 @@ include:
     - docker:dind
   variables: &deploy_stage_variables
     IMAGE_TAG: ${CI_COMMIT_SHA}
+    # IMAGE_TAG: ${CI_COMMIT_SHA}
```
Comment (douglasduteil) [inline .gitlab-ci/stages/deploy.yml]: Duplicate ?

File: .gitlab-ci/stages/deploy.yml [hunk 5]
```diff
@@ -44,87 +91,54 @@ include:
...
-  after_script:
-    # NOTE(douglasduteil): re-run env.sh
-    # Because the `after_script` runs in a separated shell context.
-    # see https://docs.gitlab.com/ee/ci/variables/where_variables_can_be_used.html#execution-shell-environment
+    VALUES_FILE: ./.k8s/api.values.yml
+  before_script:
     - source ./.gitlab-ci/env.sh
-    - echo "Api available at - ${API_URL}"
+    - HOST=${API_HOST}
+    - *resolve_helm_render_args
```
Comment (douglasduteil) [inline .gitlab-ci/stages/deploy.yml]: to remove in v3 gitlab-ci-yaml

File: .gitlab-ci/stages/deploy.yml [hunk 6]
```diff
@@ -44,87 +91,54 @@ include:
...
-    # NOTE(douglasduteil): re-run env.sh
-    # Because the `after_script` runs in a separated shell context.
-    # see https://docs.gitlab.com/ee/ci/variables/where_variables_can_be_used.html#execution-shell-environment
+    CONTEXT: frontend
+    VALUES_FILE: ./.k8s/frontend.values.yml
+  before_script:
     - source ./.gitlab-ci/env.sh
-    - echo "Frontend available at - ${FRONTEND_URL}"
+    - HOST=${FRONTEND_HOST}
+    - *resolve_helm_render_args
```
Comment (douglasduteil) [inline .gitlab-ci/stages/deploy.yml]: to remove in v3 gitlab-ci-yaml

File: .gitlab-ci/stages/deploy.yml [hunk 7]
```diff
@@ -133,12 +147,17 @@ Deploy elasticsearch (incubateur):
...
   variables:
     <<: *deploy_stage_variables
     CONTEXT: data
     ES_PORT: ${ELASTICSEARCH_PORT}
     IMAGE_NAME: ${CI_REGISTRY_IMAGE}/data
+  script:
+    - kubectl delete job ${CONTEXT} || true
+    #
+    - envsubst < ./.k8s/data.job.yml > ./job.yml
+    - kubectl apply -f ./job.yml || true
```
Comment (douglasduteil) [inline .gitlab-ci/stages/deploy.yml]: Merge both lines `envsubst < ./.k8s/data.job.yml | kubectl apply -` (je crois...)

File: .gitlab-ci/stages/deploy.yml [hunk 8]
```diff
@@ -154,28 +173,25 @@ Create Elasticsearch Index (incubateur):
...
-    # NOTE(douglasduteil): re-run env.sh
-    # Because the `after_script` runs in a separated shell context.
-    # see https://docs.gitlab.com/ee/ci/variables/where_variables_can_be_used.html#execution-shell-environment
+    CONTEXT: nlp
+    VALUES_FILE: ./.k8s/nlp.values.yml
+  before_script:
     - source ./.gitlab-ci/env.sh
-    - echo "NLP api available at - ${NLP_URL}"
+    - HOST=${NLP_HOST}
+    - *resolve_helm_render_args
```
Comment (douglasduteil) [inline .gitlab-ci/stages/deploy.yml]: to remove in v3 gitlab-ci-yaml

File: .gitlab-ci/stages/notify.yml [hunk 9]
```diff
@@ -1,89 +1,68 @@
...
+#  when: manual
+#  before_script:
+#    - apk add --no-cache git=2.22.0-r0
+#    - git checkout ${CI_COMMIT_REF_NAME}
+#    - git config user.name "${CI_GIT_AUTHOR_NAME}"
+#    - git config user.email "${CI_GIT_AUTHOR_EMAIL}"
+#    - git remote set-url origin https://${GITHUB_TOKEN}@github.com/${CI_PROJECT_PATH}.git
+#    - yarn --frozen-lockfile
+#  script:
+#    - GH_TOKEN=${GITHUB_TOKEN} yarn lerna version ${LERNA_ARGS:="--force-publish --yes"}
```
Comment (douglasduteil) [inline .gitlab-ci/stages/notify.yml]: Remove comments we have the git history ;)

File: .gitlab-ci/stages/notify.yml [hunk 10]
```diff
@@ -1,89 +1,68 @@
...
+#  except:
+#    variables:
+#      - $PRODUCTION
+#  script:
+#    - echo "Put ${CI_COMMIT_REF_NAME}@${CI_COMMIT_SHORT_SHA} in production"
+#    - curl --request POST
+#      --form ref="${CI_COMMIT_REF_NAME}"
+#      --form token="${CI_DEPLOY_TRIGGER}"
+#      --form variables[PRODUCTION]="true"
+#      https://gitlab.factory.social.gouv.fr/api/v4/projects/51/trigger/pipeline
```
Comment (douglasduteil) [inline .gitlab-ci/stages/notify.yml]: Remove comments we have the git history ;)

File: .gitlab-ci/stages/notify.yml [hunk 11]
```diff
@@ -1,89 +1,68 @@
...
+#    - git checkout ${CI_COMMIT_REF_NAME}
+#    - git config user.name "${CI_GIT_AUTHOR_NAME}"
+#    - git config user.email "${CI_GIT_AUTHOR_EMAIL}"
+#    - git remote set-url origin https://${GITHUB_TOKEN}@github.com/${CI_PROJECT_PATH}.git
+#    - yarn --frozen-lockfile
+#  script:
+#    - GH_TOKEN=${GITHUB_TOKEN} yarn lerna version ${LERNA_ARGS:="--force-publish --yes"}
 #
-
-Production:
```
Comment (douglasduteil) [inline .gitlab-ci/stages/notify.yml:74]: We need this

File: .gitlab-ci/stages/notify.yml [hunk 12]
```diff
@@ -1,89 +1,68 @@
...
-  script:
-    - echo "Notify Success to GitHub deployment id '$(cat $DEPLOY_ID_FILE)'"
-    - export URL="http://${FRONTEND_HOST}"
-    - sh ./.k8s/scripts/send-url.sh $(cat $DEPLOY_ID_FILE) success
 
-#
-#
-#
-
-Release:
```
Comment (douglasduteil) [inline .gitlab-ci/stages/notify.yml:54]: We need this

File: docker/elasticsearch/Dockerfile [hunk 13]
```diff
@@ -1,6 +1,10 @@
...
 
-RUN /usr/share/elasticsearch/bin/elasticsearch-plugin install --silent analysis-icu
+WORKDIR /usr/share/elasticsearch
 
-COPY ./elasticsearch.yml /usr/share/elasticsearch/config/elasticsearch.yml
+RUN bin/elasticsearch-plugin install --silent analysis-icu
+
+RUN  bin/elasticsearch-keystore create
+
+COPY ./elasticsearch.yml  config/elasticsearch.yml
```
Comment (douglasduteil) [inline docker/elasticsearch/Dockerfile]:
```suggestion
COPY ./elasticsearch.yml config/elasticsearch.yml
```

File: packages/code-du-travail-api/Dockerfile [hunk 14]
```diff
@@ -11,6 +11,8 @@ COPY ./package.json /app/package.json
 
 COPY --from=cdtn-base-image /app/packages/code-du-travail-api/dist /app/dist
 
+RUN yarn add elastic-apm-node --production
```
Comment (douglasduteil) [inline packages/code-du-travail-api/Dockerfile:14]: Why not a __real__ api dependency ?

File: packages/code-du-travail-api/package.json [hunk 15]
```diff
@@ -32,7 +32,7 @@
     "@elastic/elasticsearch": "7",
     "@koa/cors": "^3.0.0",
     "abort-controller": "3.0.0",
-    "dotenv": "^8.2.0",
+    "elastic-apm-node": "^3.2.0",
     "jest": "^24.9.0",
```
Comment (douglasduteil) [inline packages/code-du-travail-api/package.json]: Jest as dependencies ??

File: packages/code-du-travail-api/package.json [hunk 16]
```diff
@@ -32,7 +32,7 @@
     "@elastic/elasticsearch": "7",
     "@koa/cors": "^3.0.0",
     "abort-controller": "3.0.0",
```
Comment (douglasduteil) [inline packages/code-du-travail-api/package.json]:
```suggestion
```

Comment (douglasduteil) [reply packages/code-du-travail-api/package.json]: Require yarn.lock update

File: packages/code-du-travail-api/src/server/routes/search/index.js [hunk 17]
```diff
@@ -84,14 +84,14 @@ router.get("/search", async ctx => {
...
     );
-    const query_vector = await fetch(
-      `${NLP_URL}/api/search?q=${encodeURIComponent(query)}`
-    )
-      .then(response => (response = response.json()))
-      .catch(error => {
-        logger.error(error);
-        return [];
-      });
+    // const query_vector = await fetch(
```
Comment (douglasduteil) [inline packages/code-du-travail-api/src/server/routes/search/index.js]: We need this

**Review by douglasduteil — CHANGES_REQUESTED** [review]

File: .gitlab-ci/stages/06-deploy.yml [hunk 18]
```diff
@@ -0,0 +1,203 @@
...
+
+include:
+  #
+  - "https://raw.githubusercontent.com/SocialGouv/gitlab-ci-yml/v3.0.7/base_create_namespace_stage.yml"
+  - "https://raw.githubusercontent.com/SocialGouv/gitlab-ci-yml/v3.0.7/base_deploy_nodejs_chart_stage.yml"
+  - "https://raw.githubusercontent.com/SocialGouv/gitlab-ci-yml/v3.0.7/base_deploy_hpa_chart_stage.yml"
+  - "https://raw.githubusercontent.com/SocialGouv/gitlab-ci-yml/v3.0.7/base_docker_helm_image_stage.yml"
+  - "https://raw.githubusercontent.com/SocialGouv/gitlab-ci-yml/v3.0.7/base_docker_kubectl_image_stage.yml"
+  - "https://raw.githubusercontent.com/SocialGouv/gitlab-ci-yml/v3.0.7/base_semantic_release_stage.yml"
+  - "https://raw.githubusercontent.com/SocialGouv/gitlab-ci-yml/v3.0.7/base_register_stage.yml"
```
Comment (douglasduteil) [inline .gitlab-ci/stages/06-deploy.yml]: use 

```
- project
  file
  ref
```

File: .gitlab-ci/stages/06-deploy.yml [hunk 19]
```diff
@@ -0,0 +1,203 @@
...
+#
+
+.deploy_stage: &deploy_stage
+  stage: "Deploy"
+  dependencies: []
+  services:
+    - docker:dind
+  variables: &deploy_stage_variables
+    IMAGE_TAG: ${CI_COMMIT_SHA}
+    # IMAGE_TAG: ${CI_COMMIT_SHA}
```
Comment (douglasduteil) [inline .gitlab-ci/stages/06-deploy.yml]: to remove
```suggestion

```

File: .gitlab-ci/stages/07-notify.yml [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+  stage: Notify Finished Deployment
+  dependencies:
+    - Notify Starting Deployment
+
+Notify Success:
+  extends: .base_notify_success_stage
+  stage: Notify Finished Deployment
+  dependencies:
+    - Notify Starting Deployment
+#.notify_stage: &notify_stage
```
Comment (douglasduteil) [inline .gitlab-ci/stages/07-notify.yml]: comments to remove

**Review by douglasduteil — CHANGES_REQUESTED** [review]

**Comment by douglasduteil** [comment]
**TODO** 

- [ ] should test fast release (skip ci) on commit release
- [ ] should test production more on prod cluster

**Comment by lionelB** [comment]
@ylascombe je pense qu'il manque les secrets k8s. est ce qu'on peut se faire un point demain à ce sujet ?

**Comment by lionelB** [comment]
@douglasduteil I've updated the pipelines to take advantages of external_pull_request
- use manual deployement for feature branch (external_pull_request)
- use automatique deployement for prod
- release job only register images
- avoid running job for  release commit

File: .gitlab-ci/stages/deploy.yml [hunk 21]
```diff
@@ -133,12 +125,23 @@ Deploy elasticsearch (incubateur):
 
 .deploy-data: &deploy_data
   <<: *deploy_stage
-  extends: .k8s-deploy-data
+  image:
+    name: $CI_REGISTRY/$IMAGE_INFRA_BASE_NAME/docker-kube:latest
+    entrypoint: [""]
```
Comment (douglasduteil) [inline .gitlab-ci/stages/deploy.yml]: can be replaced by 

```
image: registry.gitlab.factory.social.gouv.fr/socialgouv/docker/kubectl:0.20.0
```

?

File: .k8s/api/deployment.yml [hunk 22]
```diff
@@ -37,23 +37,55 @@ spec:
...
+                  key: password
             - name: VERSION
               value: "${VERSION}"
             - name: NLP_URL
               value: "${NLP_API_URL}"
+            - name: ES_INDEX_PREFIX
+              value: "cdtn-${BRANCH_HASH}"
       initContainers:
         - name: wait-for-elasticsearch
+<<<<<<< HEAD
```
Comment (douglasduteil) [inline .k8s/api/deployment.yml]: Ouch !

File: .k8s/api/deployment.yml [hunk 23]
```diff
@@ -37,23 +37,55 @@ spec:
...
           image: registry.gitlab.factory.social.gouv.fr/socialgouv/docker/curl:0.14.0
+||||||| merged common ancestors
```
Comment (douglasduteil) [inline .k8s/api/deployment.yml]: Ouch !

File: .k8s/api/deployment.yml [hunk 24]
```diff
@@ -37,23 +37,55 @@ spec:
...
+          image: alpine
+=======
```
Comment (douglasduteil) [inline .k8s/api/deployment.yml]: Ouch !

File: .k8s/api/deployment.yml [hunk 25]
```diff
@@ -37,23 +37,55 @@ spec:
...
+          image: curlimages/curl:7.67.0
+>>>>>>> fix initcontainer credentials
```
Comment (douglasduteil) [inline .k8s/api/deployment.yml]: Ouch !

**Review by douglasduteil — CHANGES_REQUESTED** [review]

File: packages/code-du-travail-api/Dockerfile (hunk 14 above)
Comment (lionelB) [reply packages/code-du-travail-api/Dockerfile:14]: because it's not working with ncc

**Review by douglasduteil — APPROVED** [review]

**Pull request merged**